File: sdk-java/src/test/java/io/superdurable/dex/iwfcompat/BasicInternalChannelFlow.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.iwfcompat;

import io.superdurable.dex.Channel;
import io.superdurable.dex.ChannelMap;
import io.superdurable.dex.ConditionCombination;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepMovement;
import io.superdurable.dex.Wait;

final class BasicInternalChannelFlow implements Flow<Integer> {
    private final Channel<Integer> firstChannel =
            Channel.define("test-inter-state-channel-1", Integer.class);
    private final ChannelMap<Integer> channelMap =
            ChannelMap.define("test-inter-state-channel-map", Integer.class);
    private final ForkStep start = new ForkStep();
    private final ConsumeStep consumer = new ConsumeStep();
    private final PublishStep publisher = new PublishStep();

    @Override
    public StepList<Integer> getSteps() {
        return StepList.startStep(start).otherSteps(consumer, publisher);
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(firstChannel, channelMap);
    }

    final class ForkStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            return StepDecision.goToMulti(
                    StepMovement.of(consumer, input),
                    StepMovement.of(publisher, input));
        }
    }

    final class ConsumeStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public Wait waitFor(final Context context, final Integer input) {
            return Wait.anyCombinationOf(
                    ConditionCombination.of(firstChannel.forOne("first")),
                    ConditionCombination.of(channelMap.forOne("one")));
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            return StepDecision.gracefulComplete(
                    input + firstChannel.getConditionResults(context).size());
        }
    }

    final class PublishStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            firstChannel.publish(context, input);
            channelMap.publish(context, "one", input);
            return StepDecision.deadEnd();
        }
    }
}

File: sdk-typescript/test/iwfcompat/basic_internal_channel_flow.ts
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

import {
  Channel,
  ChannelMap,
  ConditionCombination,
  StepDef,
  StepMovement,
  Wait,
  deadEnd,
  doubleCodec,
  goToMulti,
  gracefulComplete,
  type Context,
  type Flow,
  type PersistenceSchema,
  type Step,
  type StepDecision,
} from "../../src/index.js";

class ConsumeStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public constructor(
    private readonly first: Channel<number>,
    private readonly channelMap: ChannelMap<number>,
  ) {}

  public getStepType(): string {
    return "ConsumeStep";
  }

  public waitFor(_context: Context, _input: number): Wait {
    return Wait.anyCombinationOf(
      ConditionCombination.of(this.first.forOne("first")),
      ConditionCombination.of(this.channelMap.forOne("one")),
    );
  }

  public execute(context: Context, input: number): StepDecision {
    return gracefulComplete(input + this.first.results(context).length);
  }
}

class PublishStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public constructor(
    private readonly first: Channel<number>,
    private readonly channelMap: ChannelMap<number>,
  ) {}

  public getStepType(): string {
    return "PublishStep";
  }

  public execute(context: Context, input: number): StepDecision {
    this.first.publish(context, input);
    this.channelMap.publish(context, "one", input);
    return deadEnd();
  }
}

class ForkStep implements Step<number> {
  public readonly inputCodec = doubleCodec;

  public constructor(
    private readonly consumer: ConsumeStep,
    private readonly publisher: PublishStep,
  ) {}

  public getStepType(): string {
    return "ForkStep";
  }

  public execute(_context: Context, input: number): StepDecision {
    return goToMulti(
      StepMovement.of(this.consumer, input),
      StepMovement.of(this.publisher, input),
    );
  }
}

export class BasicInternalChannelFlow implements Flow<number> {
  private readonly firstChannel = new Channel("test-inter-state-channel-1", doubleCodec);
  private readonly channelMap = new ChannelMap("test-inter-state-channel-map", doubleCodec);
  private readonly consumer = new ConsumeStep(this.firstChannel, this.channelMap);
  private readonly publisher = new PublishStep(this.firstChannel, this.channelMap);
  private readonly start = new ForkStep(this.consumer, this.publisher);

  public getFlowType(): string {
    return "BasicInternalChannelFlow";
  }

  public getSteps() {
    return [
      StepDef.startStep(this.start),
      StepDef.nonStartStep(this.consumer),
      StepDef.nonStartStep(this.publisher),
    ];
  }

  public getPersistenceSchema(): PersistenceSchema {
    return { channels: [this.firstChannel, this.channelMap] };
  }
}
